riscv/src/reachability.rs

```diff
@@ -89,11 +92,42 @@ fn extract_replacements(statements: &[Statement]) -> BTreeMap<&str, &str> {
             _ => None,
         })
         .fold(BTreeMap::new(), |mut acc, (from, to)| {
-            if acc.insert(from, to).is_some() {
+            if acc.insert(from, Cell::new(to)).is_some() {
```

Comment:
Do we really need Cell for that?

Reply:
Well, there are many ways to do it. But to mutate the tree in-place in a depth search to ensure that each transition is only walked once, this is the least ugly way to do it.

Reply:
To make this solution perfect, we just need to write a rust RFC to change `BTreeMap` interface to provide a way to convert a `BTreeMap<&str, Cell<&str>>` to a `BTreeMap<&str, &str>` without rebuilding the entire tree, as both are binary-identical.

Reply:
Why do you need to modify it in place? Wouldn't it be easier to just create a copy?

Reply:
No! Wait!

Maybe there is a way to do it if I create a Vec with the keys before doing the depth search.

Reply:
> Why do you need to modify it in place? Wouldn't it be easier to just create a copy?

Yes, but consider the case where a long path is walked to find the final name of a symbol. For every intermediate name in the path, a fraction of this same path will be walked to find the final name, which makes the algorithm O(N² log N). If I change the final name in-place as I walk the path the first time, the algorithm is then O(N log N).

Of course, this is completely irrelevant unless you expect someone to try a DoS attack on the compiler.

Reply:
Changed to the idea that collect the keys into a Vec. It is much more beautiful now. Still inplace and O(N log N).